### PR TITLE
[SPARK-40707][CONNECT] Add groupby to connect DSL and test more than one grouping expressions

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

1. Add `groupby` to connect DSL and test more than one grouping expressions
2. Pass limited data types through connect proto for LocalRelation's attributes.
3. Cleanup unused `Trait` in the testing code. 

### Why are the changes needed?

Enhance connect's support for GROUP BY.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

UT

### CODE REVIEW

**Comment by amaliujia** [comment]
R: @cloud-fan

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala [hunk 1]
```diff
@@ -51,6 +51,22 @@ package object dsl {
             .build()
         ).build()
       }
+
+      def groupBy(
+        groupingExprs: proto.Expression*)(aggregateExprs: proto.Expression*): proto.Relation = {
```
Comment (cloud-fan) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:56]:
```suggestion
          groupingExprs: proto.Expression*)(aggregateExprs: proto.Expression*): proto.Relation = {
```

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala [hunk 2]
```diff
@@ -51,6 +51,22 @@ package object dsl {
...
+        val agg = proto.Aggregate.newBuilder()
+        agg.setInput(logicalPlan)
+
+        val groupingSet = proto.Aggregate.GroupingSet.newBuilder()
+        for (groupingExpr <- groupingExprs) {
+          groupingSet.addAggregateExpressions(groupingExpr)
```
Comment (cloud-fan) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:62]: why do we add group by expression to aggregate expressions?

Comment (amaliujia) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala]: renamed the proto to make this clear.

**Comment by amaliujia** [comment]
@cloud-fan PR updated. PLAT,

File: connector/connect/src/main/protobuf/spark/connect/relations.proto [hunk 3]
```diff
@@ -129,22 +129,8 @@ message Fetch {
...
-    repeated Expression aggregate_expressions = 1;
-  }
-
-  message Measure {
-    AggregateFunction function = 1;
-    // Conditional filter for SUM(x FILTER WHERE x < 10)
-    Expression filter = 2;
-  }
+  repeated Expression grouping_expressions = 2;
+  repeated AggregateFunction result_expressions = 3;
```
Comment (amaliujia) [inline connector/connect/src/main/protobuf/spark/connect/relations.proto:133]: @cloud-fan I still keep this as `AggregateFunction`. `proto.Expression` is a too general type for now.

connect does not have a `NamedExpression`. I will follow up on this to improve this part.

This PR is to improve the grouping_expressions anyway

Comment (cloud-fan) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:134]: followup improvement SGTM. I don't think we even need `AggregateFunction`. The SQL parser usually just generate `UnresolvedFunction`, and the analyzer will look up the function and figure out if it's scalar/aggregate/window/table value function.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/TypeProtoConverter.scala [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+package org.apache.spark.sql.connect.planner
+
+import org.apache.spark.connect.proto
+import org.apache.spark.sql.types.{DataType, IntegerType, StringType}
+
+/**
+ * This object offers methods to convert to/from connect proto to catalyst types.
+ */
+object TypeProtoConverter {
+  def toCatalystType(t: proto.Type): DataType = {
```
Comment (cloud-fan) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/TypeProtoConverter.scala:27]: shall we name it `proto.DataType`? And rename this object to `DateTypeProtoConverter `

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**